File: wrappers/rust/tests/quicktx_integration_test.rs
//! Integration tests for QuickTx with Yaci DevKit.
//!
//! Requires:
//! - Yaci DevKit running on port 10000
//! - Native library built: ./gradlew :core:nativeCompile
//!
//! Run with:
//!   cd wrappers/rust && DYLD_LIBRARY_PATH=../../core/build/native/nativeCompile \
//!       cargo test --test quicktx_integration_test -- --test-threads=1

use ccl::Bridge;
use serde_json::{json, Value};
use std::thread;
use std::time::Duration;

const DEVKIT_URL: &str = "http://localhost:10000/local-cluster/api";
const DEVKIT_PROVIDER_URL: &str = "http://localhost:10000/local-cluster/api";

// --- DevKit helper functions ---

fn devkit_available() -> bool {
    ureq::get(&format!("{}/admin/devnet", DEVKIT_URL))
        .timeout(Duration::from_secs(3))
        .call()
        .map(|r| r.status() == 200)
        .unwrap_or(false)
}

fn devkit_reset() {
    let _ = ureq::post(&format!("{}/admin/devnet/reset", DEVKIT_URL))
        .timeout(Duration::from_secs(10))
        .call();
}

fn devkit_topup(address: &str, ada_amount: u64) {
    // Yaci DevKit 0.12 (companion mode) re-bootstraps the devnet on reset before handing over to the
    // node, so a topup right after reset can transiently fail. Retry with backoff.
    let body = json!({"address": address, "adaAmount": ada_amount}).to_string();
    for attempt in 1..=8 {
        let resp = ureq::post(&format!("{}/addresses/topup", DEVKIT_URL))
            .set("Content-Type", "application/json")
            .send_string(&body);
        match resp {
            Ok(r) => {
                let text = r.into_string().unwrap_or_default();
                if !text.contains("\"status\":false") {
                    return;
                }
            }
            Err(e) if attempt == 8 => panic!("topup failed after retries: {}", e),
            Err(_) => {}
        }
        thread::sleep(Duration::from_secs(4));
    }
    panic!("topup failed after retries");
}

fn devkit_get_utxos(address: &str) -> Value {
    let resp = ureq::get(&format!("{}/addresses/{}/utxos", DEVKIT_URL, address))
        .call()
        .expect("Failed to get utxos");
    resp.into_json::<Value>().expect("Invalid utxo JSON")
}

fn devkit_get_protocol_params() -> Value {
    let resp = ureq::get(&format!("{}/epochs/parameters", DEVKIT_URL))
        .call()
        .expect("Failed to get protocol params");
    resp.into_json::<Value>().expect("Invalid PP JSON")
}

fn devkit_submit_tx(tx_cbor_hex: &str) -> String {
    let tx_bytes = hex::decode(tx_cbor_hex).expect("Invalid tx hex");
    let resp = ureq::post(&format!("{}/tx/submit", DEVKIT_URL))
        .set("Content-Type", "application/cbor")
        .send_bytes(&tx_bytes)
        .expect("Failed to submit tx");

    let text = resp.into_string().expect("Failed to read response");
    text.trim().trim_matches('"').to_string()
}

fn devkit_get_tx(tx_hash: &str) -> Option<Value> {
    match ureq::get(&format!("{}/txs/{}", DEVKIT_URL, tx_hash)).call() {
        Ok(resp) => resp.into_json::<Value>().ok(),
        Err(_) => None,
    }
}

// Pull the expected treasury value out of a Conway ConwayTreasuryValueMismatch rejection, e.g.
// "... expected: Coin 43186776312112}".
fn parse_expected_treasury(submit_err: &str) -> Option<String> {
    let idx = submit_err.find("expected: Coin ")?;
    let rest = &submit_err[idx + "expected: Coin ".len()..];
    let digits: String = rest.chars().take_while(|c| c.is_ascii_digit()).collect();
    if digits.is_empty() {
        None
    } else {
        Some(digits)
    }
}

// Submit that returns the error body instead of panicking, so callers can inspect a rejection (ureq
// treats 4xx/5xx as Err by default).
fn devkit_try_submit(tx_cbor_hex: &str) -> Result<String, String> {
    let tx_bytes = hex::decode(tx_cbor_hex).map_err(|e| e.to_string())?;
    match ureq::post(&format!("{}/tx/submit", DEVKIT_URL))
        .set("Content-Type", "application/cbor")
        .send_bytes(&tx_bytes)
    {
        Ok(resp) => Ok(resp
            .into_string()
            .unwrap_or_default()
            .trim()
            .trim_matches('"')
            .to_string()),
        Err(ureq::Error::Status(code, resp)) => {
            Err(format!("submit failed ({}): {}", code, resp.into_string().unwrap_or_default()))
        }
        Err(e) => Err(e.to_string()),
    }
}

fn wait_for_block() {
    thread::sleep(Duration::from_secs(3));
}

// hex decode helper (avoid adding another dep)
mod hex {
    pub fn decode(s: &str) -> Result<Vec<u8>, String> {
        if s.len() % 2 != 0 {
            return Err("odd length".to_string());
        }
        (0..s.len())
            .step_by(2)
            .map(|i| {
                u8::from_str_radix(&s[i..i + 2], 16).map_err(|e| e.to_string())
            })
            .collect()
    }
}

fn skip_if_no_devkit() -> bool {
    if !devkit_available() {
        eprintln!("SKIP: Yaci DevKit not available on port 10000");
        return true;
    }
    false
}

fn get_testnet_account(bridge: &Bridge) -> (String, String, String) {
    let result = bridge
        .account()
        .create(ccl::network::TESTNET)
        .expect("create account");
    let json: Value = serde_json::from_str(&result).expect("parse account");
    let addr = json["base_address"].as_str().unwrap().to_string();
    let mnemonic = json["mnemonic"].as_str().unwrap().to_string();
    let stake = json["stake_address"].as_str().unwrap_or("").to_string();
    (addr, mnemonic, stake)
}

fn fund_sender(bridge: &Bridge, ada: u64) -> (String, String) {
    let (addr, mnemonic, _) = get_testnet_account(bridge);
    devkit_topup(&addr, ada);
    wait_for_block();
    (addr, mnemonic)
}

fn total_lovelace(utxos: &Value) -> u64 {
    let arr = match utxos.as_array() {
        Some(a) => a,
        None => return 0,
    };
    let mut total: u64 = 0;
    for u in arr {
        if let Some(amounts) = u["amount"].as_array() {
            for a in amounts {
                if a["unit"].as_str() == Some("lovelace") {
                    if let Some(q) = a["quantity"].as_str() {
                        total += q.parse::<u64>().unwrap_or(0);
                    } else if let Some(q) = a["quantity"].as_u64() {
                        total += q;
                    } else if let Some(q) = a["quantity"].as_f64() {
                        total += q as u64;
                    }
                }
            }
        }
    }
    total
}

// --- Integration Tests ---

fn payment_yaml(from: &str, to: &str, quantity: &str) -> String {
    format!(
        "version: 1.0\n\
         transaction:\n\
         \x20 - tx:\n\
         \x20     from: {from}\n\
         \x20     intents:\n\
         \x20       - type: payment\n\
         \x20         address: {to}\n\
         \x20         amounts:\n\
         \x20           - unit: lovelace\n\
         \x20             quantity: \"{quantity}\"\n"
    )
}

#[test]
fn test_integration_simple_ada_transfer() {
    if skip_if_no_devkit() {
        return;
    }
    devkit_reset();
    wait_for_block();

    let bridge = Bridge::new().expect("create bridge");
    let (sender, mnemonic) = fund_sender(&bridge, 150);
    let (receiver, _, _) = get_testnet_account(&bridge);

    let utxos = devkit_get_utxos(&sender);
    let pp = devkit_get_protocol_params();

    let yaml = payment_yaml(&sender, &receiver, "5000000");
    let result = bridge.quicktx().build(&yaml, &utxos, &pp, None).expect("build failed");
    assert!(!result.tx_cbor.is_empty());
    assert_eq!(result.tx_hash.len(), 64);

    let signed_tx = bridge
        .account()
        .sign_tx(&mnemonic, ccl::network::TESTNET, 0, 0, &result.tx_cbor)
        .expect("sign failed");
    let tx_hash = devkit_submit_tx(&signed_tx);
    assert!(!tx_hash.is_empty());

    wait_for_block();
    let receiver_utxos = devkit_get_utxos(&receiver);
    assert_eq!(total_lovelace(&receiver_utxos), 5_000_000);
}

#[test]
fn test_integration_insufficient_funds() {
    if skip_if_no_devkit() {
        return;
    }
    devkit_reset();
    wait_for_block();

    let bridge = Bridge::new().expect("create bridge");
    let (sender, _) = fund_sender(&bridge, 2);
    let (receiver, _, _) = get_testnet_account(&bridge);

    let utxos = devkit_get_utxos(&sender);
    let pp = devkit_get_protocol_params();

    let yaml = payment_yaml(&sender, &receiver, "100000000");
    let result = bridge.quicktx().build(&yaml, &utxos, &pp, None);
    assert!(result.is_err(), "expected insufficient funds error");
}

// The fixed test account the quicktx-intents fixtures are derived from (account 0/0).
const INTENT_MNEMONIC: &str = "test walk nut penalty hip pave soap entry language right filter choice";
const INTENT_SENDER: &str = "addr_test1qz2fxv2umyhttkxyxp8x0dlpdt3k6cwng5pxj3jhsydzer3jcu5d8ps7zex2k2xt3uqxgjqnnj83ws8lhrn648jjxtwq2ytjqp";

// Treasury donation: Conway validates the tx's declared current_treasury_value against the node's
// live treasury. Learn the required value from the ledger's own rejection — submit, read the
// expected value out of a ConwayTreasuryValueMismatch, rebuild with it, and resubmit (retrying also
// absorbs an epoch boundary landing between attempts).
#[test]
fn test_integration_donation_treasury() {
    if skip_if_no_devkit() {
        return;
    }
    devkit_reset();
    wait_for_block();
    devkit_topup(INTENT_SENDER, 6000);
    wait_for_block();

    let bridge = Bridge::new().expect("create bridge");
    let utxos = devkit_get_utxos(INTENT_SENDER);
    let pp = devkit_get_protocol_params();
    let base_yaml = std::fs::read_to_string("../../test-fixtures/quicktx-intents/donation.yaml")
        .expect("read donation fixture");

    let mut treasury = String::from("0");
    let mut last_err = String::new();
    for _ in 0..5 {
        let yaml = base_yaml.replace(
            "current_treasury_value: 0",
            &format!("current_treasury_value: {}", treasury),
        );
        let result = bridge.quicktx().build(&yaml, &utxos, &pp, None).expect("build");
        let signed = bridge
            .account()
            .sign_tx(INTENT_MNEMONIC, ccl::network::TESTNET, 0, 0, &result.tx_cbor)
            .expect("sign");
        match devkit_try_submit(&signed) {
            Ok(tx_hash) => {
                assert!(!tx_hash.is_empty(), "empty tx hash from submit");
                return; // accepted
            }
            Err(e) => {
                last_err = e;
                match parse_expected_treasury(&last_err) {
                    Some(v) => treasury = v,
                    None => panic!("unexpected submit error: {}", last_err),
                }
            }
        }
    }
    panic!("donation submit failed after retries: {}", last_err);
}

File: wrappers/go/ccl/quicktx_integration_test.go
package ccl

import (
	"bufio"
	"bytes"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"strings"
	"testing"
	"time"
)

// Yaci DevKit integration tests for QuickTx.
//
// Requires:
// - Yaci DevKit running on port 10000
// - Native library built: ./gradlew :core:nativeCompile
//
// Run with:
//   cd wrappers/go && go test -v -run Integration ./ccl/

const devkitURL = "http://localhost:10000/local-cluster/api"
const devkitProviderURL = "http://localhost:10000/local-cluster/api"

// devkitAvailable checks if Yaci DevKit is running.
func devkitAvailable() bool {
	client := &http.Client{Timeout: 3 * time.Second}
	resp, err := client.Get(devkitURL + "/admin/devnet")
	if err != nil {
		return false
	}
	defer resp.Body.Close()
	return resp.StatusCode == 200
}

func devkitReset() {
	req, _ := http.NewRequest("POST", devkitURL+"/admin/devnet/reset", nil)
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err == nil {
		resp.Body.Close()
	}
}

func devkitTopup(address string, adaAmount int) error {
	body, _ := json.Marshal(map[string]interface{}{
		"address":   address,
		"adaAmount": adaAmount,
	})
	resp, err := http.Post(devkitURL+"/addresses/topup", "application/json", bytes.NewReader(body))
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		b, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("topup failed (%d): %s", resp.StatusCode, string(b))
	}
	return nil
}

func devkitGetUtxos(address string) ([]map[string]interface{}, error) {
	resp, err := http.Get(devkitURL + "/addresses/" + address + "/utxos")
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var utxos []map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&utxos); err != nil {
		return nil, err
	}
	return utxos, nil
}

func devkitGetProtocolParams() (map[string]interface{}, error) {
	resp, err := http.Get(devkitURL + "/epochs/parameters")
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var pp map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&pp); err != nil {
		return nil, err
	}
	return pp, nil
}

func devkitSubmitTx(txCborHex string) (string, error) {
	// Use raw TCP to avoid Go's strict "duplicate chunked TE" rejection
	txBytes, err := hex.DecodeString(txCborHex)
	if err != nil {
		return "", fmt.Errorf("invalid tx hex: %w", err)
	}

	conn, err := net.DialTimeout("tcp", "localhost:10000", 10*time.Second)
	if err != nil {
		return "", fmt.Errorf("connect: %w", err)
	}
	defer conn.Close()
	conn.SetDeadline(time.Now().Add(30 * time.Second))

	// Write raw HTTP request
	var reqBuf bytes.Buffer
	fmt.Fprintf(&reqBuf, "POST /local-cluster/api/tx/submit HTTP/1.1\r\n")
	fmt.Fprintf(&reqBuf, "Host: localhost:10000\r\n")
	fmt.Fprintf(&reqBuf, "Content-Type: application/cbor\r\n")
	fmt.Fprintf(&reqBuf, "Content-Length: %d\r\n", len(txBytes))
	fmt.Fprintf(&reqBuf, "Connection: close\r\n")
	fmt.Fprintf(&reqBuf, "\r\n")
	reqBuf.Write(txBytes)

	if _, err := conn.Write(reqBuf.Bytes()); err != nil {
		return "", fmt.Errorf("write: %w", err)
	}

	// Read response
	reader := bufio.NewReader(conn)
	// Read status line
	statusLine, err := reader.ReadString('\n')
	if err != nil {
		return "", fmt.Errorf("read status: %w", err)
	}
	parts := strings.SplitN(strings.TrimSpace(statusLine), " ", 3)
	if len(parts) < 2 {
		return "", fmt.Errorf("malformed status: %s", statusLine)
	}

	// Read headers until empty line
	var contentLength int
	for {
		line, err := reader.ReadString('\n')
		if err != nil {
			return "", fmt.Errorf("read header: %w", err)
		}
		line = strings.TrimSpace(line)
		if line == "" {
			break
		}
		if strings.HasPrefix(strings.ToLower(line), "content-length:") {
			fmt.Sscanf(strings.TrimSpace(line[15:]), "%d", &contentLength)
		}
	}

	// Read body
	var body []byte
	if contentLength > 0 {
		body = make([]byte, contentLength)
		_, err = io.ReadFull(reader, body)
	} else {
		body, err = io.ReadAll(reader)
	}
	if err != nil && err != io.EOF {
		return "", fmt.Errorf("read body: %w", err)
	}

	statusCode := parts[1]
	if statusCode != "200" && statusCode != "202" {
		return "", fmt.Errorf("submit failed (%s): %s", statusCode, string(body))
	}

	txHash := strings.TrimSpace(string(body))
	txHash = strings.Trim(txHash, "\"")
	return txHash, nil
}

func devkitGetTx(txHash string) (map[string]interface{}, error) {
	resp, err := http.Get(devkitURL + "/txs/" + txHash)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("get tx failed (%d)", resp.StatusCode)
	}
	var tx map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&tx); err != nil {
		return nil, err
	}
	return tx, nil
}

func waitForBlock() {
	time.Sleep(3 * time.Second)
}

func skipIfNoDevKit(t *testing.T) {
	t.Helper()
	if !devkitAvailable() {
		t.Skip("Yaci DevKit not available on port 10000")
	}
}

func fundSender(t *testing.T, ada int) *AccountInfo {
	t.Helper()
	account, err := bridge.Account.Create(Testnet)
	if err != nil {
		t.Fatalf("create account: %v", err)
	}
	if err := devkitTopup(account.BaseAddress, ada); err != nil {
		t.Fatalf("topup: %v", err)
	}
	waitForBlock()
	return account
}

func totalLovelace(utxos []map[string]interface{}) int64 {
	var total int64
	for _, u := range utxos {
		amounts, ok := u["amount"].([]interface{})
		if !ok {
			continue
		}
		for _, a := range amounts {
			am, ok := a.(map[string]interface{})
			if !ok {
				continue
			}
			if am["unit"] == "lovelace" {
				switch q := am["quantity"].(type) {
				case float64:
					total += int64(q)
				case string:
					var v int64
					fmt.Sscanf(q, "%d", &v)
					total += v
				}
			}
		}
	}
	return total
}

// --- Integration Tests ---

func TestIntegrationSimpleADATransfer(t *testing.T) {
	skipIfNoDevKit(t)
	devkitReset()
	waitForBlock()

	sender := fundSender(t, 150)
	receiver, _ := bridge.Account.Create(Testnet)

	utxos, err := devkitGetUtxos(sender.BaseAddress)
	if err != nil {
		t.Fatalf("get utxos: %v", err)
	}
	pp, err := devkitGetProtocolParams()
	if err != nil {
		t.Fatalf("get pp: %v", err)
	}

	yaml := quickTxYaml(sender.BaseAddress, receiver.BaseAddress, "5000000")
	result, err := bridge.QuickTx.Build(yaml, utxos, pp)
	if err != nil {
		t.Fatalf("build: %v", err)
	}
	assertTxResult(t, result)

	signedTx, err := bridge.Account.SignTx(sender.Mnemonic, Testnet, 0, 0, result.TxCbor)
	if err != nil {
		t.Fatalf("sign: %v", err)
	}
	txHash, err := devkitSubmitTx(signedTx)
	if err != nil {
		t.Fatalf("submit: %v", err)
	}
	if txHash == "" {
		t.Fatal("empty tx hash from submit")
	}

	waitForBlock()
	receiverUtxos, err := devkitGetUtxos(receiver.BaseAddress)
	if err != nil {
		t.Fatalf("get receiver utxos: %v", err)
	}
	if total := totalLovelace(receiverUtxos); total != 5_000_000 {
		t.Errorf("expected 5 ADA (5000000), got %d lovelace", total)
	}
}

func TestIntegrationMultipleReceivers(t *testing.T) {
	skipIfNoDevKit(t)
	devkitReset()
	waitForBlock()

	sender := fundSender(t, 150)
	r1, _ := bridge.Account.Create(Testnet)
	r2, _ := bridge.Account.Create(Testnet)

	utxos, err := devkitGetUtxos(sender.BaseAddress)
	if err != nil {
		t.Fatalf("get utxos: %v", err)
	}
	pp, err := devkitGetProtocolParams()
	if err != nil {
		t.Fatalf("get pp: %v", err)
	}

	yaml := fmt.Sprintf(`
version: 1.0
transaction:
  - tx:
      from: %s
      intents:
        - type: payment
          address: %s
          amounts:
            - unit: lovelace
              quantity: "3000000"
        - type: payment
          address: %s
          amounts:
            - unit: lovelace
              quantity: "2000000"
`, sender.BaseAddress, r1.BaseAddress, r2.BaseAddress)

	result, err := bridge.QuickTx.Build(yaml, utxos, pp)
	if err != nil {
		t.Fatalf("build: %v", err)
	}
	signedTx, _ := bridge.Account.SignTx(sender.Mnemonic, Testnet, 0, 0, result.TxCbor)
	if _, err := devkitSubmitTx(signedTx); err != nil {
		t.Fatalf("submit: %v", err)
	}

	waitForBlock()
	if total := totalLovelace(mustUtxos(t, r1.BaseAddress)); total != 3_000_000 {
		t.Errorf("expected 3 ADA for r1, got %d", total)
	}
}

func TestIntegrationInsufficientFunds(t *testing.T) {
	skipIfNoDevKit(t)
	devkitReset()
	waitForBlock()

	sender := fundSender(t, 2)
	receiver, _ := bridge.Account.Create(Testnet)

	utxos, _ := devkitGetUtxos(sender.BaseAddress)
	pp, _ := devkitGetProtocolParams()

	yaml := quickTxYaml(sender.BaseAddress, receiver.BaseAddress, "100000000")
	if _, err := bridge.QuickTx.Build(yaml, utxos, pp); err == nil {
		t.Fatal("expected insufficient funds error")
	}
}

func mustUtxos(t *testing.T, address string) []map[string]interface{} {
	t.Helper()
	utxos, err := devkitGetUtxos(address)
	if err != nil {
		t.Fatalf("get utxos: %v", err)
	}
	return utxos
}
